himbaechel/uarch/gatemate/route_clock.cc

```diff
@@ -0,0 +1,162 @@
+
+    auto feeds_clk_port = [](PortRef &port) {
+        return port.cell->type.in(id_CPE_HALF, id_CPE_HALF_L, id_CPE_HALF_U) && port.port.in(id_CLK);
+    };
+
+    auto pip_plane = [&](PipId pip) {
+        const auto &extra_data =
+                *reinterpret_cast<const GateMatePipExtraDataPOD *>(chip_pip_info(ctx->chip_info, pip).extra_data.get());
+        if (extra_data.type != PipExtra::PIP_EXTRA_MUX)
+            return uint8_t{0};
+        return extra_data.plane;
+    };
+
+    for (auto &net : ctx->nets) {
+        NetInfo *glb_net = net.second.get();
+        if (!glb_net->driver.cell || glb_net->driver.cell->type != id_BUFG)
+            continue;
+
+        for (auto &usr : glb_net->users) {
+            if (feeds_clk_port(usr)) {
+                clk_nets.push_back(glb_net);
+                break;
+            }
+        }
+    }
+
+    log_info("Routing clock nets...\n");
+    for (auto glb_net : clk_nets) {
+        log_info("    routing net '%s'\n", glb_net->name.c_str(ctx));
+        ctx->bindWire(ctx->getNetinfoSourceWire(glb_net), glb_net, STRENGTH_LOCKED);
+
+        auto bufg_idx = ctx->getBelLocation(glb_net->driver.cell->bel).z;
+
+        for (auto &usr : glb_net->users) {
+            std::priority_queue<QueuedWire, std::vector<QueuedWire>, std::greater<QueuedWire>> visit;
+            dict<WireId, PipId> backtrace;
+            WireId dest = WireId();
+            // skip arcs that are not part of lowskew routing
+            if (!feeds_clk_port(usr))
+                continue;
+
+            auto cpe_loc = ctx->getBelLocation(usr.cell->bel);
+
+            auto sink_wire = ctx->getNetinfoSinkWire(glb_net, usr, 0);
+            if (ctx->debug) {
+                auto sink_wire_name = "(uninitialized)";
+                if (sink_wire != WireId())
+                    sink_wire_name = ctx->nameOfWire(sink_wire);
+                log_info("        routing arc to %s.%s (wire %s):\n", usr.cell->name.c_str(ctx), usr.port.c_str(ctx),
+                         sink_wire_name);
+            }
+            visit.push(QueuedWire(sink_wire));
+            while (!visit.empty()) {
+                QueuedWire curr = visit.top();
+                visit.pop();
+                if (curr.wire == ctx->getNetinfoSourceWire(glb_net)) {
+                    if (ctx->debug)
+                        log_info("            (%.3fns)\n", curr.delay);
+                    dest = curr.wire;
+                    break;
+                }
+
+                PipId bound_pip;
+                auto fnd_wire = glb_net->wires.find(curr.wire);
+                if (fnd_wire != glb_net->wires.end()) {
+                    bound_pip = fnd_wire->second.pip;
+                }
+
+                for (auto uh : ctx->getPipsUphill(curr.wire)) {
+                    if (!ctx->checkPipAvailForNet(uh, glb_net))
+                        continue;
+                    WireId src = ctx->getPipSrcWire(uh);
+                    if (backtrace.count(src))
```

Comment:
this means you won't do any revisits if you subsequently find a cheaper route to a wire - is this intentional?

Reply:
Because there's no lookahead, this router is essentially Dijkstra, so I'm pretty sure the first route we find is the cheapest.

Reply:
ah right, I missed that, thanks for explaining